pyteal/ast/opup.py

```diff
@@ -0,0 +1,157 @@
+                requires the app_id to target for inner app calls. Defaults
+                to None.
+        """
+
+        # With only OnCall and Explicit modes supported, the mode argument
+        # isn't strictly necessary but it will most likely be required if
+        # we do decide to add more modes in the future.
+        if mode == OpUpMode.Explicit:
+            if target_app_id is None:
+                raise TealInputError(
+                    "target_app_id must be specified in Explicit OpUp mode"
+                )
+            require_type(target_app_id, TealType.uint64)
+            self.target_app_id = target_app_id
+        elif mode == OpUpMode.OnCall:
+            if target_app_id is not None:
+                raise TealInputError("target_app_id is not used in OnCall OpUp mode")
+        else:
+            raise TealInputError("Invalid OpUp mode provided")
+
+        self.mode = mode
+
+    def _construct_itxn(self) -> Expr:
+        if self.mode == OpUpMode.Explicit:
+            return Seq(
+                InnerTxnBuilder.Begin(),
+                InnerTxnBuilder.SetFields(
+                    {
+                        TxnField.type_enum: TxnType.ApplicationCall,
+                        TxnField.application_id: self.target_app_id,
+                    }
+                ),
+                InnerTxnBuilder.Submit(),
+            )
+        else:
+            return Seq(
+                InnerTxnBuilder.Begin(),
+                InnerTxnBuilder.SetFields(
+                    {
+                        TxnField.type_enum: TxnType.ApplicationCall,
+                        TxnField.on_completion: OnComplete.DeleteApplication,
+                        TxnField.approval_program: ON_CALL_APP,
+                        TxnField.clear_state_program: ON_CALL_APP,
+                    }
+                ),
+                InnerTxnBuilder.Submit(),
+            )
+
+    def ensure_budget(self, required_budget: Expr) -> Expr:
+        """Ensure that the budget will be at least the required_budget.
+
+        Note: the available budget just prior to calling ensure_budget() must be
+        high enough to execute the budget increase code. The exact budget required
+        depends on the provided required_budget expression, but a budget of ~20
+        should be sufficient for most use cases. If lack of budget is an issue then
+        consider moving the call to ensure_budget() earlier in the pyteal program."""
+        require_type(required_budget, TealType.uint64)
+
+        # A budget buffer is necessary to deal with an edge case of ensure_budget():
+        #   if the current budget is equal to or only slightly higher than the
+        #   required budget then it's possible for ensure_budget() to return with a
+        #   current budget less than the required budget. The buffer prevents this
+        #   from being the case.
+        buffer = Int(10)
+        buffered_budget = ScratchVar(TealType.uint64)
```

Comment:
just one question: is these lines of code ensuring that in the case of: *current budget slightly higher than required budget*, the `_construct_itxn` will be executed at least once?

Since I am not sure if I understand
> it's possible for ensure_budget() to return with a current budget less than the required budget.

Reply:
Say the user calls the method with `required_budget=1000` then, after issuing an inner app call, the current budget is 1000. When the comparison `current_budget < required_budget` is executed it will return false, but the comparison itself consumes budget so the resulting budget may be something like 999 instead of 1000.